[FLINK-20153] Add documentation for BATCH execution mode (#14114)
apache/flink: docs/dev/datastream_execution_mode.md

execution mode when your program is unbounded because only this mode is general
enough to be able to deal with continuous data streams.

One obvious outlier case is when you want to use a bounded job to bootstrap
some job state that you then want to use in an unbounded job. For example, by
running a bounded job using STREAMING mode, taking a savepoint, and then
restoring that savepoint on an unbounded job. This is a very specific use case
and one that might soon become obsolete when we allow producing a savepoint as
additional output of a BATCH execution job.

Another case where you might run a bounded job using STREAMING mode is when
writing tests for code that will eventually run with unbounded sources. For
testing it can be more natural to use a bounded source in those cases.

## Configuring BATCH execution mode

The execution mode can be configured via the `execution.runtime-mode` setting.
There are three possible values:

 - `STREAMING`: The classic DataStream execution mode
 - `BATCH`: Batch-style execution on the DataStream API
 - `AUTOMATIC`: Let the system decide based on the boundedness of the sources

>  This can be configured either in `flink-conf.yaml`, via command line

Author: You're right, I'll take out the mention of `flink-conf.yaml`. 👌

 parameters of `bin/flink run ...`, or programmatically when
 creating/configuring the `StreamExecutionEnvironment`.

 Here's how you can configure the execution mode via the command line:

 ```bash
 $ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
 ```

 This example shows how you can configure the execution mode in code:

 ```java
Configuration config = new Configuration();
config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
 ```

## Execution Behavior

This section provides an overview of the execution behavior of BATCH execution
mode and contrasts it with STREAMING execution mode. For more details, please
refer to the FLIPs that introduced this feature:

> [FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and

Author: I click the "Share" button/menu at the top and take the link from there. I'm hoping that this is stable, because just taking a URL like https://cwiki.apache.org/confluence/display/FLINK/FLIP-134%3A+Batch+execution+for+the+DataStream+API is not stable when someone renames the title. There's also the option of using https://s.apache.org/FLIP-134, which is a shortlink I manually created.

run all the tasks at the same time.

Network shuffles are _pipelined_, meaning that records are immediately sent to
downstream tasks, with some buffering on the network layer. Again, this enables
low-latency processing.

#### BATCH Execution Mode

In BATCH execution mode, the tasks of a job can be separated into stages that
can be executed one after another. We can do this because the input is bounded
and Flink can therefore fully process one stage of the pipeline before moving
on to the next. In the above example the job would have three stages that
correspond to the three tasks that are separated by the shuffle barriers.

Instead of sending records immediately to downstream tasks, as explained above
for STREAMING mode, processing in stages requires Flink to materialize
intermediate results of tasks to some non-ephemeral storage which allows
downstream tasks to read them after upstream tasks have already gone off line.
This will increase the latency of processing but comes with other interesting
properties. For one, this allows Flink to backtrack to the latest available
results when a failure happens instead of restarting the whole job. Another

> side effect is that BATCH jobs can execute on fewer resources (in terms of

Reviewer: ```suggestion
side effect is that `BATCH` jobs can execute on fewer resources (in terms of
```

available slots at TaskManagers) because the system can execute tasks
sequentially one after the other.

TaskManagers will keep intermediate results at least as long as downstream
tasks have not consumed them. After that, they will be kept for as long as

> space allows in order to allow the aforementioned backtracking to earlier

Reviewer: I think technically we keep them until all consuming pipelined regions have produced their outputs.

Author: I added an explanation in parentheses but right now it will point to the non-existing documentation about pipelined regions. That's another TODO. 😅

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

The DataStream API supports different runtime execution modes from which you
can choose depending on the requirements of your use case and the
characteristics of your job.

There is the "classic" execution behavior of the DataStream API which we call
`STREAMING` execution mode. This should be used for unbounded jobs that require
continuous incremental processing and are expected to stay online indefinitely.

Additionally, there is a batch-style execution mode that we call `BATCH`
execution mode. This executes jobs in a way that is more reminiscent of batch
processing frameworks such as MapReduce. This should be used for bounded jobs
for which you have a known fixed input and which do not run continuously.

We have these different execution modes because `BATCH` execution allows some

> additional optimizations that we can only do when we now that our input is

Reviewer 2: ```suggestion
additional optimizations that we can only do when we know that our input is
```

bounded. For example, different join/aggregation strategies can be used, in

> additional to a different shuffle implementation that allows more efficient

Reviewer 3: ```suggestion
addition to a different shuffle implementation that allows more efficient
```

failure recovery behavior. We will go into some of the details of the execution
behavior below.

* This will be replaced by the TOC
{:toc}

## When can/should I use BATCH execution mode?

The BATCH execution mode can only be used for Jobs/Flink Programs that are
_bounded_. Boundedness is a property of a data source that tells us whether all
the input coming from that source is known before execution or whether new data

> will show up, potentially for forever. A job, in turn, is bounded if all its

Reviewer 2: ```suggestion
will show up, potentially indefinitely. A job, in turn, is bounded if all its
```

> sources are bounded and unbounded otherwise.

Reviewer 2: ```suggestion
sources are bounded, and unbounded otherwise.
```

STREAMING execution mode, on the other hand, can be used for both bounded and
unbounded jobs.

Another term that is typically used for bounded sources is _batch source_. Or,
we can say that we are working with a batch data set or a batch of data. Some

> typical other terms for an unbounded source are _continuous source_, _streaming

Reviewer 2: ```suggestion other terms for an unbounded source are _continuous source_, _streaming
```

Reviewer 3: Shall we just drop the paragraph whatsoever? Let's maybe establish a single term for it?

Author: Ok, I'm dropping it. Also fine for you, @alpinegizmo ?

Reviewer 2: 👍 for dropping it

Reviewer 4: Drop it

source_, _stream_, or _infinite stream_.

As a rule of thumb, you should be using BATCH execution mode when your program
is bounded because this will be more efficient. You have to use STREAMING
execution mode when your program is unbounded because only this mode is general
enough to be able to deal with continuous data streams.

> TODO: Should we even go into this?

Reviewer 2: I'm in favor of keeping at least some mention of the bootstrapping case. We could also mention that it makes sense to write tests that run in STREAMING mode against bounded sources.

Reviewer 3: +1 for outlining the testing case

Author: Will do

One obvious outlier case is when you want to use a bounded job to bootstrap
some job state that you then want to use in an unbounded job. For example, by

> running a bounded job using STREAMING mode, taking a savepoint and then

Reviewer 2: ```suggestion
running a bounded job using STREAMING mode, taking a savepoint, and then
```

restoring that savepoint on an unbounded job. This is a very specific use case
and one that might soon become obsolete when we allow producing a savepoint as
additional output of a BATCH execution job.

## Configuring BATCH execution mode

The execution mode can be configured via the `execution.runtime-mode` setting.
There are three possible values:

 - `STREAMING`: The classic DataStream execution mode
 - `BATCH`: Batch-style execution on the DataStream API
 - `AUTOMATIC`: Let the system decide based on the boundedness of the sources

>  This can be configured either in the `flink-conf.yaml`, via command line

Reviewer 2: ```suggestion
 This can be configured either in `flink-conf.yaml`, via command line
```

 Here's how you can configure the execution mode via the command line:

 ```bash
 $ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
 ```

 This example shows how you can configure the execution mode in code:

 ```java
Configuration config = new Configuration();
config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
 ```

## Execution Behavior

This section provides an overview of the execution behavior of BATCH execution
mode and contrasts it with STREAMING execution mode. For more details, please
refer to the FLIPs that introduced this feature:
[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the
documentation about [task scheduling (TODO)](<TODO>).

### Task Scheduling And Network Shuffle

Flink jobs consist of different operations that are connected together in a
dataflow graph. The system decides how to schedule the execution of these
operations on different processes/machines (TaskManager) and how data is

> shuffled (sent) betwixt them.

Reviewer 3: I appreciate learning a new word `betwixt`, but I think it is nicer towards our readers, which are very often not native speakers, to stick to the simpler `between`.

Author: I just put it there for someone to find it. 😊

Reviewer 2: I was bemused to encounter such a fine word, and correctly assumed I didn't need to be the one to react to it. 😏

Author: 🧐

Multiple operations/operators can be chained together using a feature called
[chaining]({% link dev/stream/operators/index.md
%}#task-chaining-and-resource-groups). A group of one or multiple (chained)
operators that Flink considers as a unit of scheduling is called a _task_.
Often the term _subtask_ is used to refer to the individual instances of tasks
that are running in parallel on multiple TaskManagers but we will only use the
term _task_ here.

> Task scheduling and the network shuffle work different for BATCH execution mode

Reviewer 2: ```suggestion
Task scheduling and the network shuffle work differently for BATCH execution mode
```

and STREAMING execution mode. Mostly due to the fact that we know our input
data is bounded in BATCH execution mode, which allows Flink to use more
efficient data structures and algorithms.

We will use this example to explain the differences in task scheduling and
network transfer:

```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

DataStreamSource<String> source = env.fromElements(...);

source.name("source")
	.map(...).name("map1")
	.map(...).name("map2")
	.rebalance()
	.map(...).name("map3")
	.map(...).name("map4")
	.keyBy((value) -> value)
	.map(...).name("map5")
	.map(...).name("map6")

> 	.sinkTo(...).name("sink");

Reviewer 2: A diagram of an execution graph for this topology would be helpful.

Author: I was afraid someone would suggest this, but yes, I assumed that I would have to do this... 😅🙈

Author: I added a figure.

`map()`, `flatMap()`, or `filter()` can just forward data straight to the next
operation, which allows these operations to be chained together. This means
that Flink would not normally insert a network shuffle between them.

Operation such as `keyBy()` or `rebalance()` on the other hand require data to
be shuffled between different parallel instances of tasks. This induces a
network shuffle.

For the above example Flink would group operations together as tasks like this:

- Task1: `source`, `map1`, and `map2`
- Task2: `map3`, `map4`
- Task3: `map5`, `map6`, and `sink`

And we have a network shuffle between Tasks 1 and 2, and also Tasks 2 and 3.

#### STREAMING Execution Mode

In STREAMING execution mode, all tasks need to be online/running all the time.
This allows Flink to immediately process new records through the whole
pipeline, which we need for low-latency stream processing. This also means that
the TaskManagers that are allotted to a job need to have enough resources to
run all the tasks at the same time.

Network shuffles are _pipelined_, meaning that records are immediately sent to
downstream tasks, with some buffering on the network layer. Again, this enables

> low-latency processing.

Reviewer 3: It's good the way it is now. Just me nitpicking.

I'd say it is not (only) about low-latency. It's (also) because of the unbounded characteristic of the processing. There is no point in the processing that we can say e.g. an operator will not see any more records and we can drop its state. We must keep its state on all the time because it might change at any point in time in the future.

Reviewer 4: I agree, I think continuous is more important to highlight here than latency

Author: I expanded this paragraph. PTAL.

#### BATCH Execution Mode

In BATCH execution mode, the tasks of a job can be separated into stages that
can be executed one after another. We can do this because the input is bounded
and Flink can therefore fully process one stage of the pipeline before moving
on to the next. In the above example the job would have three stages that
correspond to the three tasks that are separated by the shuffle barriers.

Instead of sending records immediately to downstream tasks, as explained above
for STREAMING mode, processing in stages requires Flink to materialize
intermediate results of tasks to some non-ephemeral storage which allows
downstream tasks to read them after upstream tasks have already gone off line.
This will increase the latency of processing but comes with other interesting
properties. For one, this allows Flink to backtrack to the latest available

> results when a failure happens instead of restarting the whole job.

Reviewer 3: Shall we also mention we can run the job with less resources than the sum of all operators required resources?

Author: Will do

TaskManagers will keep intermediate results at least as long as downstream
tasks have not consumed them. After that, they will be kept for as long as
space allows in order to allow the aforementioned backtracking to earlier
results in case of a failure.

### State Backends / State

- mention that in BATCH mode a special state backend is used and not the
  configured one

### Event Time / Watermarks

### Processing Time

### Failure Recovery

> In STREAMING execution mode, Flink uses checkpoints for failure recover. Take a

Reviewer 2: ```suggestion
In STREAMING execution mode, Flink uses checkpoints for failure recovery. Take a
```

---

> title: "Execution Mode (Batch/Streaming)"

Reviewer 4: Can you add "Execution Mode" to the glossary? 

Also, does it make sense to add a note to the `dev/batch/index.md` directing new applications to use bounded datastream or table? Or are we not there yet?

Author: I pushed a commit that adds it and I'm also recommending to use the DataStream API in the DataSet docs. I think it's time for a bold move... 😅

nav-id: datastream_execution_mode
nav-parent_id: streaming
nav-pos: 1
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

The DataStream API supports different runtime execution modes from which you
can choose depending on the requirements of your use case and the
characteristics of your job.

> There is the "classic" execution behavior of the DataStream API which we call

Reviewer 4: ```suggestion
There is the "classic" execution behavior of the DataStream API, which we call
```

`STREAMING` execution mode. This should be used for unbounded jobs that require
continuous incremental processing and are expected to stay online indefinitely.

Additionally, there is a batch-style execution mode that we call `BATCH`
execution mode. This executes jobs in a way that is more reminiscent of batch
processing frameworks such as MapReduce. This should be used for bounded jobs
for which you have a known fixed input and which do not run continuously.

> We have these different execution modes because `BATCH` execution allows some
> additional optimizations that we can only do when we know that our input is
> bounded. For example, different join/aggregation strategies can be used, in
> addition to a different shuffle implementation that allows more efficient
> failure recovery behavior. We will go into some of the details of the execution
> behavior below.

Reviewer 4: ```suggestion
Apache Flink's unified approach to stream and batch processing means that a DataStream application executed
over bounded input will produce the same results regardless of the configured execution mode.
By enabling `BATCH` execution, we allow Flink to apply additional optimizations that we can only do work when we know that our input is bounded. For example, different join/aggregation strategies can be used, in
addition to a different shuffle implementation that allows more efficient
failure recovery behavior. We will go into some of the details of the execution
behavior below.
```

* This will be replaced by the TOC
{:toc}

## When can/should I use BATCH execution mode?

> The BATCH execution mode can only be used for Jobs/Flink Programs that are

Reviewer 4: ```suggestion
The `BATCH` execution mode can only be used for Jobs/Flink Programs that are
```

_bounded_. Boundedness is a property of a data source that tells us whether all
the input coming from that source is known before execution or whether new data
will show up, potentially indefinitely. A job, in turn, is bounded if all its
sources are bounded, and unbounded otherwise.

> STREAMING execution mode, on the other hand, can be used for both bounded and

Reviewer 4: ```suggestion
`STREAMING` execution mode, on the other hand, can be used for both bounded and
```

unbounded jobs.

Another term that is typically used for bounded sources is _batch source_. Or,
we can say that we are working with a batch data set or a batch of data. Some
typical other terms for an unbounded source are _continuous source_, _streaming
source_, _stream_, or _infinite stream_.

> As a rule of thumb, you should be using BATCH execution mode when your program

Reviewer 4: ```suggestion
As a rule of thumb, you should be using `BATCH` execution mode when your program
```

> is bounded because this will be more efficient. You have to use STREAMING

Reviewer 4: ```suggestion
is bounded because this will be more efficient. You have to use `STREAMING`
```

execution mode when your program is unbounded because only this mode is general
enough to be able to deal with continuous data streams.

One obvious outlier case is when you want to use a bounded job to bootstrap
some job state that you then want to use in an unbounded job. For example, by

> running a bounded job using STREAMING mode, taking a savepoint, and then

Reviewer 4: ```suggestion
running a bounded job using `STREAMING` mode, taking a savepoint, and then
```

restoring that savepoint on an unbounded job. This is a very specific use case
and one that might soon become obsolete when we allow producing a savepoint as

> additional output of a BATCH execution job.

Reviewer 4: ```suggestion
additional output of a `BATCH` execution job.
```

> Another case where you might run a bounded job using STREAMING mode is when

Reviewer 4: ```suggestion
Another case where you might run a bounded job using `STREAMING` mode is when
```

writing tests for code that will eventually run with unbounded sources. For
testing it can be more natural to use a bounded source in those cases.

## Configuring BATCH execution mode

The execution mode can be configured via the `execution.runtime-mode` setting.
There are three possible values:

>  - `STREAMING`: The classic DataStream execution mode

Reviewer 4: ```suggestion
 - `STREAMING`: The classic DataStream execution mode (default)
```

I'm assuming `STREAMING` is the default. If not, please mark the correct one.

 - `BATCH`: Batch-style execution on the DataStream API
 - `AUTOMATIC`: Let the system decide based on the boundedness of the sources

 This can be configured either in `flink-conf.yaml`, via command line
 parameters of `bin/flink run ...`, or programmatically when
 creating/configuring the `StreamExecutionEnvironment`.

 Here's how you can configure the execution mode via the command line:

 ```bash
 $ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
 ```

 This example shows how you can configure the execution mode in code:

 ```java
Configuration config = new Configuration();
config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
 ```

## Execution Behavior

This section provides an overview of the execution behavior of BATCH execution
mode and contrasts it with STREAMING execution mode. For more details, please
refer to the FLIPs that introduced this feature:
[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the

> documentation about [task scheduling (TODO)](<TODO>).

Reviewer 4: reminder for too

network transfer:

```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

DataStreamSource<String> source = env.fromElements(...);

source.name("source")
	.map(...).name("map1")
	.map(...).name("map2")
	.rebalance()
	.map(...).name("map3")
	.map(...).name("map4")
	.keyBy((value) -> value)
	.map(...).name("map5")
	.map(...).name("map6")
	.sinkTo(...).name("sink");
```

Operations that imply a 1-to-1 connection pattern between operations, such as
`map()`, `flatMap()`, or `filter()` can just forward data straight to the next
operation, which allows these operations to be chained together. This means
that Flink would not normally insert a network shuffle between them.

Operation such as `keyBy()` or `rebalance()` on the other hand require data to
be shuffled between different parallel instances of tasks. This induces a
network shuffle.

For the above example Flink would group operations together as tasks like this:

- Task1: `source`, `map1`, and `map2`
- Task2: `map3`, `map4`
- Task3: `map5`, `map6`, and `sink`

And we have a network shuffle between Tasks 1 and 2, and also Tasks 2 and 3.

#### STREAMING Execution Mode

> In STREAMING execution mode, all tasks need to be online/running all the time.

Reviewer 4: ```suggestion
In `STREAMING` execution mode, all tasks need to be online/running all the time.
```

This allows Flink to immediately process new records through the whole

> pipeline, which we need for low-latency stream processing. This also means that

Reviewer 4: ```suggestion
pipeline, which we need for continuous and low-latency stream processing. This also means that
```
?

Instead of sending records immediately to downstream tasks, as explained above
for STREAMING mode, processing in stages requires Flink to materialize
intermediate results of tasks to some non-ephemeral storage which allows
downstream tasks to read them after upstream tasks have already gone off line.
This will increase the latency of processing but comes with other interesting
properties. For one, this allows Flink to backtrack to the latest available
results when a failure happens instead of restarting the whole job. Another
side effect is that BATCH jobs can execute on fewer resources (in terms of
available slots at TaskManagers) because the system can execute tasks
sequentially one after the other.

TaskManagers will keep intermediate results at least as long as downstream
tasks have not consumed them. After that, they will be kept for as long as
space allows in order to allow the aforementioned backtracking to earlier
results in case of a failure.

### State Backends / State

- mention that in BATCH mode a special state backend is used and not the
  configured one

### Event Time / Watermarks

### Processing Time

### Failure Recovery

> In STREAMING execution mode, Flink uses checkpoints for failure recovery. Take a

Reviewer 4: ```suggestion
In `STREAMING` execution mode, Flink uses checkpoints for failure recovery. Take a
```

look at the [checkpointing documentation]({% link
dev/stream/state/checkpointing.md %}) for hands-on documentation about this and
how to configure it. There is also a more introductory section about [fault
tolerance via state snapshots]({% link learn-flink/fault_tolerance.md %}) that
explains the concepts at a higher level.

One of the characteristics of checkpointing for failure recovery is that Flink
will restart all the running tasks from a checkpoint in case of a failure. This

> can be more costly than what we have to do in BATCH mode (as explained below),

Reviewer 4: ```suggestion
can be more costly than what we have to do in `BATCH` mode (as explained below),
```

> which is one of the reasons that you should use BATCH execution mode if your

Reviewer 4: ```suggestion
which is one of the reasons that you should use `BATCH` execution mode if your
```

job allows it.

> In BATCH execution mode, Flink will try and backtrack to previous processing

Reviewer 4: ```suggestion
In `BATCH` execution mode, Flink will try and backtrack to previous processing
```

stages for which intermediate results are still available. Potentially, only
the tasks that failed (or their predecessors in the graph) will have to be
restarted, which can improve processing efficiency and overall processing time
of the job compared to restarting all tasks from a checkpoint.

## Important Considerations

> What doesn't work:
>  - broadcast state/pattern
>  - iterations
>  - operations that rely on checkpointing
>  - this includes most "regular" sinks

Reviewer 4: ```suggestion
Full unification of batch and stream processing under the DataStream API is still under active development. 
That means there are still a number of DataStream features that only work under the `STREAMING` execution mode. 

 - broadcast state/pattern
 - iterations
 - operations that rely on checkpointing
 - this includes most "regular" sinks
```

Aljoscha, I'll need you to expand on the bullet points.

Author: Sorry, I forgot to mention that this documentation is WIP and that Dawid has some incoming changes for this part. 🙈

Author: Sorry, Teach

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

The DataStream API supports different runtime execution modes from which you
can choose depending on the requirements of your use case and the
characteristics of your job.

There is the "classic" execution behavior of the DataStream API, which we call
`STREAMING` execution mode. This should be used for unbounded jobs that require
continuous incremental processing and are expected to stay online indefinitely.

Additionally, there is a batch-style execution mode that we call `BATCH`
execution mode. This executes jobs in a way that is more reminiscent of batch
processing frameworks such as MapReduce. This should be used for bounded jobs
for which you have a known fixed input and which do not run continuously.

Apache Flink's unified approach to stream and batch processing means that a
DataStream application executed over bounded input will produce the same

> results regardless of the configured execution mode.  By enabling `BATCH`

Reviewer 5: remove space in the "  By enabling" (it has two spaces).

Author: 👍

execution, we allow Flink to apply additional optimizations that we can only do

> work when we know that our input is bounded. For example, different

Reviewer 5: remove the "work": "to apply additional optimizations that we can only do
work when we know that our input is bounded" -> "to apply additional optimizations that we can only do
when we know that our input is bounded"

Author: 🆗

join/aggregation strategies can be used, in addition to a different shuffle

> implementation that allows more efficient failure recovery behavior. We will go

Reviewer 5: "that allows more efficient failure recovery behavior" -> "that allows more efficient scheduling and failure recovery."

Author: 👌

into some of the details of the execution behavior below.

* This will be replaced by the TOC
{:toc}

## When can/should I use BATCH execution mode?

The `BATCH` execution mode can only be used for Jobs/Flink Programs that are
_bounded_. Boundedness is a property of a data source that tells us whether all
the input coming from that source is known before execution or whether new data
will show up, potentially indefinitely. A job, in turn, is bounded if all its
sources are bounded, and unbounded otherwise.

`STREAMING` execution mode, on the other hand, can be used for both bounded and
unbounded jobs.

As a rule of thumb, you should be using `BATCH` execution mode when your program
is bounded because this will be more efficient. You have to use `STREAMING`
execution mode when your program is unbounded because only this mode is general

> enough to be able to deal with continuous data streams.

Reviewer: Here it sounds as if it does not really matter whether to choose `BATCH` or `STREAMING` for a bounded job from a correctness perspective. However, the `FileSink` won't commit the in-progress files at the end of the program when using the `STREAMING` execution mode. It might be worthwhile to document this behaviour somewhere.

Author: Yes, this is unfortunate. Though the fact that we cannot do checkpoints as soon as at least one task has finished, which in turn means that we can't get a "final" checkpoint has been a feature/bug of DataStream execution since the beginning. I wouldn't document it here but we can think about adding this to a general "caveats" section. I'm sure there would be other corner cases that are worth documenting 😅

> One obvious outlier case is when you want to use a bounded job to bootstrap

Reviewer 5: "One obvious outlier case is when you want" -> "One obvious outlier is when you want"

Author: fixing

Another case where you might run a bounded job using `STREAMING` mode is when
writing tests for code that will eventually run with unbounded sources. For
testing it can be more natural to use a bounded source in those cases.

## Configuring BATCH execution mode

The execution mode can be configured via the `execution.runtime-mode` setting.
There are three possible values:

 - `STREAMING`: The classic DataStream execution mode (default)
 - `BATCH`: Batch-style execution on the DataStream API
 - `AUTOMATIC`: Let the system decide based on the boundedness of the sources

This can be configured via command line parameters of `bin/flink run ...`, or
programmatically when creating/configuring the `StreamExecutionEnvironment`.

Here's how you can configure the execution mode via the command line:

```bash
$ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
```

This example shows how you can configure the execution mode in code:

 ```java
Configuration config = new Configuration();
config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
 ```

> 

Reviewer 5: I would add: 

```
or 

```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
env.setRuntimeMode(RuntimeExecutionMode.BATCH);
 ```

<span class="label label-danger">Important Note</span>: We recommend users to NOT use this method
but set the `execution.runtime-mode` using the command-line when submitting the application. Keeping 
the application code configuration-free allows for more flexibility as the same application will be
able to be executed in any execution mode.

```

Author: Thanks, that's a good suggestion!

## Execution Behavior

This section provides an overview of the execution behavior of `BATCH`
execution mode and contrasts it with `STREAMING` execution mode. For more
details, please refer to the FLIPs that introduced this feature:
[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the
documentation about [task scheduling (TODO)](<TODO>).

### Task Scheduling And Network Shuffle

Flink jobs consist of different operations that are connected together in a
dataflow graph. The system decides how to schedule the execution of these

> operations on different processes/machines (TaskManager) and how data is

Reviewer 5: TaskManager -> TaskManagers

Author: fixing

shuffled (sent) between them.

Multiple operations/operators can be chained together using a feature called
[chaining]({% link dev/stream/operators/index.md
%}#task-chaining-and-resource-groups). A group of one or multiple (chained)
operators that Flink considers as a unit of scheduling is called a _task_.
Often the term _subtask_ is used to refer to the individual instances of tasks
that are running in parallel on multiple TaskManagers but we will only use the
term _task_ here.

> Task scheduling and the network shuffle work differently for `BATCH` execution

Reviewer 5: "Task scheduling and the network shuffle" -> "Task scheduling and network shuffles"

"differently for `BATCH` execution mode and `STREAMING` execution mode" -> "differently for `BATCH` and `STREAMING` execution mode"

Author: fixing

can be executed one after another. We can do this because the input is bounded
and Flink can therefore fully process one stage of the pipeline before moving
on to the next. In the above example the job would have three stages that
correspond to the three tasks that are separated by the shuffle barriers.

Instead of sending records immediately to downstream tasks, as explained above
for `STREAMING` mode, processing in stages requires Flink to materialize
intermediate results of tasks to some non-ephemeral storage which allows
downstream tasks to read them after upstream tasks have already gone off line.
This will increase the latency of processing but comes with other interesting
properties. For one, this allows Flink to backtrack to the latest available
results when a failure happens instead of restarting the whole job. Another
side effect is that `BATCH` jobs can execute on fewer resources (in terms of
available slots at TaskManagers) because the system can execute tasks
sequentially one after the other.

TaskManagers will keep intermediate results at least as long as downstream
tasks have not consumed them. (Technically, they will be kept until the

> consuming [pipelined regions TODO](<TODO>) have produced their output in turn.)

Reviewer 5: I would remove the "in turn".

Author: fixing

of the job compared to restarting all tasks from a checkpoint.

## Important Considerations

Compared to classic `STREAMING` execution mode, in `BATCH` mode some things
might not work as expected. Some features will work slightly differently while
others are not supported.

Behavior Change in BATCH mode:

* "Rolling" operations such as [reduce()]({% link dev/stream/operators/index.md
  %}#reduce) or [sum()]({% link dev/stream/operators/index.md %}#aggregations)
  emit an incremental update for every new record that arrives in `STREAMING`
  mode. In `BATCH` mode, these operations are not "rolling". They emit only the
  final result.

Unsupported in BATCH mode:

* [Checkpointing]({% link concepts/stateful-stream-processing.md
  %}#stateful-stream-processing) and any operations that depend on
  checkpointing do not work.
* [Broadcast State]({% link dev/stream/state/broadcast_state.md %})
* [Iterations]({% link dev/stream/operators/index.md %}#iterate)

Custom operators should be implemented with care, otherwise they might behave
improperly. See also additional explanations below for more details.

### Checkpointing

> As explained [above](#failure-recovery), fault tolerance for batch programs

Reviewer 5: "fault tolerance for batch programs does not use checkpointing" -> "failure recovery for batch programs
does not use checkpointing". 

I would suggest that because technically it is not fault tolerance as much as leveraging caching to speed up the catch up phase of reprocessing.

Author: good catch!

does not use checkpointing. Recovery happens by fully replaying the streams.
This is possible because inputs are bounded.  This pushes the cost more towards
the recovery, but makes the regular processing cheaper, because it avoids

> checkpoints.

Reviewer: The last paragraph of `failure recovery` reads as if `BATCH` execution improves the overall execution time of jobs but here it reads a bit differently. Concretely that `BATCH` recoveries are more costly than `STREAMING` recoveries.

Author: Good catch! @dawidwys what was the original intention here?

Reviewer 3: My intention was to briefly remind the batch failure recovery model. For that I actually reused the description from: https://ci.apache.org/projects/flink/flink-docs-master/concepts/stateful-stream-processing.html#state-and-fault-tolerance-in-batch-programs

Reviewer 3: With the description in the `failure recovery` section, we can probably drop the first paragraph and start with the second one:

>It is important to remember that because there are no checkpoints, as described [above](#failure-recovery), certain ...

BTW shall we update it in `#stateful-stream-processing.html#state-and-fault-tolerance-in-batch-programs`? It is not easy to tell which model recovers "faster" as it very much depends on the state size, number of records to replay, number of tasks to recover etc.

Author: Oh boy, this text was first added in 2016: https://github.com/apache/flink/blame/b04d51a129a3341887e7a0866557c9871f58e94c/docs/concepts/concepts.md. I copied it to the current concepts section from there. That's not at all up-to-date anymore.

That section needs an overhaul or should be removed because it's also misleading for `DataSet` programs or Table/SQL batch programs.

Author: For this documentation here I think we can go with @dawidwys's suggestion and just drop that paragraph because I added some text about that above.

%}#rolling-policy) won't work. If you need a transactional sink that works in
`BATCH` mode make sure it uses the Unified Sink API as proposed in
[FLIP-143](https://cwiki.apache.org/confluence/x/KEJ4CQ).

You can still use all the [state primitives]({% link dev/stream/state/state.md
%}#working-with-state), it's just that the mechanism used for failure recovery
will be different.

### Broadcast State

The feature was introduced to allow users to implement use-cases where a
“control” stream needs to be broadcast to all downstream tasks, and the
broadcast elements, e.g. rules, need to be applied to all incoming elements
from another stream.

In this pattern, Flink provides no guarantees about the order in which the
inputs are read.  Use-cases like the one above make sense in the streaming
world where jobs are expected to run for a long period with input data that are
not known in advance. In these settings, requirements may change over time
depending on the incoming data.

In the batch world though, we believe that such use-cases do not make much
sense, as the input (both the elements and the control stream) are static and

> known in advance.

Author: I'm afraid so. But it's not that nice. We do want to add proper support for broadcast input in the next release, though.
